incremental.go

```diff
@@ -196,3 +198,168 @@ func (p *IncrementalProducer) commitManifest(m *manifest, wopt *WriterOptions) e
 	}
 	return writer.Commit()
 }
+
+// ------------------------------------------------------------------------------------------
+
+// IncrementalConsumeFunc is a parsing callback which is run by the consumer every sync interval.
+type IncrementalConsumeFunc func(*ReaderIter) (data interface{}, err error)
+
+// IncrementalConsumerOptions configure the consumer instance.
+type IncrementalConsumerOptions struct {
+	ConsumerOptions
+}
+
+func (o *IncrementalConsumerOptions) norm() {
+	if o.Interval == 0 {
+		o.Interval = time.Minute
+	}
+}
+
+// NewIncrementalConsumer starts a new feed consumer.
+func NewIncrementalConsumer(ctx context.Context, bucketURL string, opt *IncrementalConsumerOptions, icfn IncrementalConsumeFunc) (Consumer, error) {
+	bucket, err := bfs.Connect(ctx, bucketURL)
+	if err != nil {
+		return nil, err
+	}
+
+	csm, err := NewIncrementalConsumerForBucket(ctx, bucket, opt, icfn)
+	if err != nil {
+		_ = bucket.Close()
+		return nil, err
+	}
+	csm.(*incrementalConsumer).ownBucket = true
+	return csm, nil
+}
+
+// NewIncrementalConsumerForBucket starts a new feed consumer with a remote.
+func NewIncrementalConsumerForBucket(ctx context.Context, bucket bfs.Bucket, opt *IncrementalConsumerOptions, icfn IncrementalConsumeFunc) (Consumer, error) {
+	var o IncrementalConsumerOptions
+	if opt != nil {
+		o = IncrementalConsumerOptions(*opt)
+	}
+	o.norm()
+
+	ctx, stop := context.WithCancel(ctx)
+	c := &incrementalConsumer{
+		bucket:   bucket,
+		manifest: bfs.NewObjectFromBucket(bucket, "manifest.json"),
+		opt:      o,
+		ctx:      ctx,
+		stop:     stop,
+		icfn:     icfn,
+	}
+
+	// run initial sync
+	if _, err := c.sync(true); err != nil {
+		_ = c.Close()
+		return nil, err
+	}
+
+	// start continuous loop
+	go c.loop()
+
+	return c, nil
+}
+
+type incrementalConsumer struct {
+	manifest  *bfs.Object
+	bucket    bfs.Bucket
+	ownBucket bool
+
+	opt  IncrementalConsumerOptions
+	ctx  context.Context
+	stop context.CancelFunc
+
+	icfn IncrementalConsumeFunc
+
+	consumerState
+}
+
+// Close implements Consumer interface.
+func (c *incrementalConsumer) Close() (err error) {
+	c.stop()
+	if e := c.manifest.Close(); e != nil {
+		err = e
+	}
+	if c.ownBucket {
+		if e := c.bucket.Close(); e != nil {
+			err = e
+		}
+	}
+	return
+}
+
+func (c *incrementalConsumer) sync(force bool) (*ConsumerSync, error) {
+	start := time.Now()
+	defer func() {
+		c.consumerState.updateLastSync(start)
+	}()
+
+	// retrieve original last modified time
+	lastMod, err := remoteLastModified(c.ctx, c.manifest)
+	if err != nil {
+		return nil, err
+	}
+
+	// skip update if not forced or modified
+	if !force && lastMod > 0 && lastMod.Millis() == atomic.LoadInt64(&c.lastMod) {
+		return &ConsumerSync{Consumer: c}, nil
+	}
+
+	// fetch remote manifest
+	manifest, err := loadManifest(c.ctx, c.manifest)
+	if err != nil {
+		return nil, err
+	}
+
+	// sort file list
+	files := manifest.Files
+	sort.Strings(files) // TODO! allow setting sort function in options?
```

Comment:
this is just a placeholder really....  I dont fully understand how the Generation field will be used.  will there be files of multiple generations being looped over?  will it be fine just to loop over files in order [generation ASC date ASC] and dedupe by some key (decided by caller within IncrementalConsumeFunc)?

Reply:
I assume (and am working as if) its something like
Gen 0: we produce incremental files -> `[data-0-11111.pbz, data-0-22222.pbz, data-0-33333.pbz]` 
Gen 1: we compact oldest files and delete them -> `[data-1-22222.pbz, data-0-33333.pbz. ]` (or does every file generation get bumped?)...  
Gen 1: we then produce a new incremental -> `[data-1-22222.pbz, data-0-33333.pbz, data-1-55555.pbz]`

I think `[data-0-11111.pbz, data-0-22222.pbz] -> [data-1-22222.pbz]` is correct (filename timestamp is always the last modified date @ Generation 0) so that we know the compacted data is "older" than subsequent new incremental files.

Reply:
so in ^^ case, we would ignore generation when sorting files, just loop over in timestamp order.

Reply:
the other assumption I'm making on compaction (unrelated to this MR again...)  is that our compact logic doesnt care about the type of data stored, it'll just compare rows of bytes and dedupe somehow (although it wont be fully duplicate records as by definition the record will be updated between files)

Reply:
Many thoughts here, most of them right though. Let's not worry about compaction and only look at `Files` for now. you shouldn't need to sort them here, they should be sorted by the producer!!

Once you do compaction, yes, we will start with `gen: 0, files: [data-0-111.pbz data-0-222.pbz data-0-333.pbz]` and after compaction, we will end up with `gen: 1, files: [data-1-333.pbz]`, after a few production cycles we will eventually get to `gen: 1, files: [data-1-333.pbz data-1-444.pbz data-1-555.pbz]`, then another compaction will result in `gen: 2, files: [data-2-555.pbz]` (and we can delete gen-0 files). And yes, we don't care about the content, we assume that files are streams can be concatenated (once unzipped).